kotlin/android/app/src/main/java/dev/firezone/android/tunnel/TunnelService.kt

```diff
@@ -209,17 +211,27 @@ class TunnelService : VpnService() {
     }
 
     private fun onSessionDisconnected(error: String?) {
+        Log.d(TAG, "onSessionDisconnected: $error")
         sessionPtr = null
         onTunnelStateUpdate(Tunnel.State.Down)
 
-        if (shouldReconnect && error == null) {
+        // At the moment, the FFI swallows the error.
+        // We check if it's 401 by checking if the error is null.
+        val is401Unauthorized = error != null
```

Comment:
Unfortunately this was a bug in this tunnel implementation. `error` is never null because the JVM doesn't support nullable types, so `null` actually gets stringified to `"null"`.